keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
```

Comment:
Above in the doc I see

> only allowing CEL expressions that pass compilation and type checking significantly reduces the opportunities for runtime errors.

If the schema for the config is located elsewhere (CRD I suspect), how will this be done?  I might be ok not doing it, but I'd like to understand the limitations of what we're providing.

Reply:
The basic idea (which gets complicated when we dig into details) is to do something to CRD registration where each API server watches all validatingAdmissionPolicy resources and for each policy resource update, reads the configuration CRD and type checks against it.


There can be races. CR validation  has a very similar race. Special case: CRD doesn't exist yet. For that one we'd need to fail type checking until the CRD becomes available.

Reporting type check errors is problematic. We can put them in the status, but since this is an API server configuration object, we don't have a leader elected controller so it's tricky to write to status.

Reply:
This was a challenge for Gatekeeper as well but we relied on the status field of constraints and constraint templates. 

> We can put them in the status, but since this is an API server configuration object, we don't have a leader elected controller so it's tricky to write to status.

Does this mean this is still TBD?

Reply:
Yes, I'd characterize this part of the design as incomplete or TBD. I _really_ want to offer typesafety in this design, to whatever extent is possible. But it hinges on working out design details like this.

Reply:
We've added a "Writing to Status" section that I hope resolves this.